Blog post: Comparison lantern vs pgvector

rzavalet opened the pull request:
Hi! This is the PR for the lantern vs pgvector blog post. 

The google docs draft can be found here: https://docs.google.com/document/d/1HTFod9QcGocGgpeyhQpP7144r1WebfdB48pkSEGZzsg/edit?usp=sharing

Additional material:
- Modified benchmark: https://github.com/binidxaba/ann-benchmarks
- Complete results: https://github.com/binidxaba/lantern-pgvector-comparison

Any additional feedback is appreciated!!

samay-sharma commented:
@rzavalet : Is it good for me to take a pass to review? Or are there comments from the google doc still left to resolve?

rzavalet commented:
> @rzavalet : Is it good for me to take a pass to review? Or are there comments from the google doc still left to resolve?

Yes! Please have a look when you get a chance. This PR addresses all feedback in the google doc. 

Let me know if there is anything else that can be improved.

ChuckHend commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,153 @@
+
+![Baseline throughput](./004-baseline-qps.png)
+
+Ok, the recall is similar, however the resulting QPS for Lantern is worse when compared to Pgvector. Pgvector can process between 72% and 81% more queries per second.
```
just a suggestion, but maybe say "Pgvector outperforms Lantern in QPS", or something similar. Feels more positive in tone.

rzavalet replied:
Agreed. I incorporated this change.

ChuckHend commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,153 @@
+
+![Recall](./015-recall-m24.png)
+
+![Throughput](./016-qps-m24.png)
```
Could add a note between the recall and throughput graphs, something like "pgvector has higher throughput in all values of ef_search"

rzavalet replied:
Done!

ChuckHend commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,153 @@
+- Build Time
+- Index Size
+- Recall
+- Throughput
```
Do you happen to have latency results as well? That would also be interesting to see IMO.

rzavalet replied:
The benckmark gets percentiles. I incorporated p95 plots.

ChuckHend commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,153 @@
+---
+slug: postgres-vector-search-pgvector-and-lantern
+title: 'Comparing Postgres Vector Search approaches: Pgvector vs Lantern'
```
```suggestion
title: 'Benchmarking Postgres Vector Search approaches: Pgvector vs Lantern'
```

Thoughts on calling out Benchmarking in the title? I'd say this is a benchmarking blog :)

rzavalet replied:
I changed it to "Benchmarking". @samay-sharma wdyt?

samay-sharma commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,162 @@
+title: 'Benchmarking Postgres Vector Search approaches: Pgvector vs Lantern'
+authors: [rjzv]
+tags: [postgres, extensions, vector, lantern, pgvector, hnsw]
+image: ./elephant-with-lantern.jpeg
```
Can you work with Abby to have a better on-brand image?

rzavalet replied:
Done!

samay-sharma commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,162 @@
+Image credit: Generated with Bing Image Creator
+</em></p>
+
+Today, let's talk about Vector Search in Postgres from yet another perspective. In this blog post, we'll cover:
```
We should have a better start to the post.

Vector search in Postgres is a space that has seen very active development in the last few months. While pgvector is known to most people, a few weeks ago we came across Lantern which also builds a Postgres based vector database. So, we thought about benchmarking both to compare the two approaches. In this post, we'll cover:

rzavalet replied:
Done

samay-sharma commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,162 @@
+- Benchmarks comparing index creation time, size, throughput and recall.
+- Summary of the results
+
+## Vector Search in PostgreSQL
```
Intro to Vector Search in PostgreSQL

samay-sharma replied:
Let's also avoid repetition based on what I've asked to change in the intro.

rzavalet replied:
Done

samay-sharma commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,162 @@
+
+![Baseline p95](./005-baseline-p95.png)
+
+Ok, the recall is similar, however Pgvector outperforms Lantern in QPS. Specifically, Pgvector can process between 72% and 81% more queries per second.
```
Let's also change the text to incorporate latency comparison.

rzavalet replied:
Done

samay-sharma commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,162 @@
+![Recall](./013-recall-m16.png)
+
+
+And again, Pgvector has better throughput and latencies.
```
can we say X% better throughput and Y% better latencies.

rzavalet replied:
Done

samay-sharma commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,162 @@
+
+## Conclusions
+
+Pgvector is the most popular Postgres extension for vector search. At the time of this writing, the github repository counts 7.2K stars and is actively being discussed on the web. It is also supported on most managed Postgres providers (including [Tembo Cloud](https://tembo.io/)) so is easier for you to access.
```
Replace link with cloud.tembo.io

rzavalet replied:
Done

samay-sharma reviewed (commented):
Mostly looks good. I primarily had feedback on the intro and a few other minor points. I think it'll be good to go after addressing these.

samay-sharma commented on docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,162 @@
+
+But Pgvector is not the only one in this space. The community has come up with other alternatives as well. One such example was [pg_embedding](https://github.com/neondatabase/pg_embedding). In July, [Neon](https://neon.tech/) published a [post](https://neon.tech/blog/pg-embedding-extension-for-vector-search) where they showed how its extension was 20x faster than Pgvector by using HNSW indexes .  ...
+
+Another Postgres extension in the vector search domain is [Lantern](https://github.com/lanterndata/lantern), developed by a [company of the same name](https://lantern.dev/). In October, a post on their blog claimed that their extension could outperform Pgvector by 90x in index creation time. That was an outstanding achievement!
```
If we're referring to the blog post, let's also link to it.

rzavalet replied:
Done!

samay-sharma commented on docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,157 @@
+
+For a quantitative comparison, I took the [ANN-Benchmarks](https://github.com/erikbern/ann-benchmarks) and [extended it](https://github.com/binidxaba/ann-benchmarks) to support the Lantern extension. It was just a matter of creating a copy of the Pgvector directory, naming it Lantern and making a few adjustments to use the corresponding API.
+
+The benchmark starts a container that has Postgres 14 installed and then it enables the corresponding extension. Then, it inserts a dataset into a table and builds a vector index (in both cases an HNSW index). After that, it executes a bunch of queries and evaluates the recall.
```
It says Postgres 15 at the bottom and Postgres 14 here. Which one is correct?

binidxaba replied:
It is postgres 15. Fixed.

samay-sharma commented on docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,157 @@
+
+- Quick background on doing vector search in PostgreSQL.
+- Comparing Pgvector and Lantern in terms of syntax and ease of use.
+- Benchmarks comparing index creation time, size, throughput, and recall.
```
Let's put latency here too.

binidxaba replied:
Done

samay-sharma commented on docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,157 @@
+- Benchmarks comparing index creation time, size, throughput, and recall.
+- Summary of the results
+
+## Intro to Vector Search in PostgreSQL
```
Let's see if there are any opportunities to link to your vector related blog posts in this section. Let's not write about them but just find opportunities to link to them.

binidxaba replied:
Added a couple of links.

samay-sharma commented on docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,157 @@
+
+
+## Conclusions
+
```
Can we add a table summary of results here which makes it easy to grasp the results of the blog? It's just that we share so many data points in this section that I forget what the numbers were like.

binidxaba replied:
Done, let me know what you think.

rzavalet replied:
(Sorry, I was using another account for my replies)

samay-sharma replied:
looks good

samay-sharma commented on docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:
```diff
@@ -0,0 +1,157 @@
+
+Lantern is a young project that leverages the popular [USearch engine](https://github.com/unum-cloud/usearch/). As of today, the extension has been starred more than 400 times in github and is in very active development. 
+
+Both extensions offer similar API and support the HNSW index. According to my experiments (Pgvector 0.5.1 and Lantern 0.0.11), Lantern's index creation is faster and produces smaller indexes. However, Pgvector provides better recall and throughput.
```
Comment on latency too.

binidxaba replied:
Done

samay-sharma reviewed (commented):
What I believe should be the last round of feedback. Once you have the image and have addressed these, I'll merge.
